File: src/main/java/com/cyhone/classpath/ClassPath.java
package com.cyhone.classpath;

import com.cyhone.exception.EnvCheckError;
import com.cyhone.utils.FilePathUtils;
import com.sun.javafx.scene.shape.PathUtils;
import org.apache.commons.io.FileUtils;
import org.apache.commons.lang3.StringUtils;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;

/**
 * @author cyhone
 * @date 2017/2/14
 */
public class ClassPath {
    private Entry bootClasspath;
    private Entry extClasspath;
    private Entry userClasspath;

    public static Entry newEntry(String path) throws FileNotFoundException {
        if (path.contains(File.pathSeparator)) {
            return new CompositeEntry(path);
        }

        if (path.endsWith("*")) {
            return new WildcardEntry(path);
        }

        if (path.endsWith(".jar") || path.endsWith(".JAR")
                || path.endsWith(".zip") || path.endsWith(".ZIP")) {
            return new ZipJarEntry(path);
        }
        return new DirEntry(path);
    }

    public ClassPath() {
    }

    public ClassPath(String jreOPtion, String cpOption) {
        try {
            parse(jreOPtion, cpOption);
        } catch (FileNotFoundException e) {
            throw new EnvCheckError(e.getMessage());
        }

    }

    public void parse(String jreOPtion, String cpOption) throws FileNotFoundException {
        parseBootAndExtClasspath(jreOPtion);
        parseUserClasspath(cpOption);
    }

    private void parseBootAndExtClasspath(String cpOption) throws FileNotFoundException {
        String jreDir = getJreDir(cpOption);

        //  jre/lib/*
        String jreLibPath = FilePathUtils.join(jreDir, "lib", "*");
        this.bootClasspath = new WildcardEntry(jreLibPath);

        // jre/lib/ext/*
        String jreExtPath = FilePathUtils.join(jreDir, "lib", "ext", "*");
        this.extClasspath = new WildcardEntry(jreExtPath);
    }

    private void parseUserClasspath(String cpOption) throws FileNotFoundException {
        if (StringUtils.isEmpty(cpOption)) {
            cpOption = ".";
        }
        this.userClasspath = newEntry(cpOption);
    }

    private String getJreDir(String cpOption) {
        if (!StringUtils.isEmpty(cpOption) && isExists(cpOption)) {
            return cpOption;
        }

        if (isExists("./jre")) {
            return "./jre";
        }
        String JAVA_HOME = System.getenv("JAVA_HOME");
        if (JAVA_HOME != null) {
            return FilePathUtils.join(JAVA_HOME, "jre");
        }
        throw new EnvCheckError("Can not find jre folder!");
    }

    private boolean isExists(String filepath) {
        return new File(filepath).exists();
    }

    public byte[] readClass(String className) {
        className = className + ".class";
        byte[] classBytes = null;

        try {
            classBytes = bootClasspath.readClass(className);
        } catch (IOException e) {
        }

        try {
            classBytes = extClasspath.readClass(className);
        } catch (IOException e1) {
        }

        try {
            classBytes = userClasspath.readClass(className);
        } catch (IOException e2) {
        }

        return classBytes;
    }
}

File: src/classpath/class_path.cpp
//
// Created by cyhone on 18-2-10.
//
#include <toyjvm/classpath/class_path.h>
#include <toyjvm/common/ioutils.h>
#include <toyjvm/common/system.h>
#include <toyjvm/common/strutils.h>

namespace jvm {
    static std::string GetJreDir(const std::string &jre_option);

    ClassPath::ClassPath(const std::string &jre_option, const std::string &cp_option)
    {
        ParseBootAndExtClassPath(jre_option);
        ParseUserClassPath(cp_option);
    }

    bytes ClassPath::ReadClass(const std::string &class_name)
    {
        auto cname = class_name + ".class";
        bytes data;
        try {
            data = boot_path_->ReadClass(cname);
        } catch (...) {
            try {
                data = ext_path_->ReadClass(cname);
            } catch (...) {
                data = user_path_->ReadClass(cname);
            }
        }

        return data;
    }

    void ClassPath::ParseBootAndExtClassPath(const std::string &jre_option)
    {
        auto jre_dir = GetJreDir(jre_option);

        boot_path_ = new WildCardPathEntry(jre_dir + "/lib/*");
        ext_path_ = new WildCardPathEntry(jre_dir + "/lib/ext/*");
    }

    void ClassPath::ParseUserClassPath(const std::string &class_path)
    {
        user_path_ = PathEntryFactory(class_path.empty() ? "." : class_path);
    }

    std::string GetJreDir(const std::string &jre_option)
    {
        if (!jre_option.empty() && FileExists(jre_option)) {
            return jre_option;
        }

        if (FileExists("./jre")) {
            return "./jre";
        }

        std::string java_home = GetEnv("JAVA_HOME");

        if (!java_home.empty()) {
            return java_home + "/" + "jre";
        }

        throw "can not find jre folder";
    }
}
